### PR TITLE
Fix: Add missing playbook run capability in livenet images

### PR DESCRIPTION
**Pull request by oxedions**
Adding missing diskless images customizing / playbook running capabilty.

Still a draft

Missing documentation in the readme for now. In next commit.

### CODE REVIEW

File: roles/addons/diskless/tasks/main.yml [hunk 1]
```diff
@@ -31,6 +38,9 @@
    - /var/www/html/preboot_execution_environment/diskless
    - /var/www/html/preboot_execution_environment/diskless/images
    - /var/www/html/preboot_execution_environment/diskless/kernels
+   - /etc/bluebanquise/tools/disklessset/
+   - /etc/bluebanquise/tools/disklessset/inventory/
```
Comment (oxedions) [inline roles/addons/diskless/tasks/main.yml:42]: Forgot to remove these ones. Will do in next commit.

**Comment by btravouillon** [comment]
FYI, resize an image with XFS from 2300MiB to 3GiB:

```
# unsquashfs /var/www/html/preboot_execution_environment/diskless/images/selinux4/squashfs.img

# ls -lh squashfs-root/LiveOS/rootfs.img
-rw-r--r--. 1 root root 2.3G Aug 14 19:44 squashfs-root/LiveOS/rootfs.img

# stat squashfs-root/LiveOS/rootfs.img
  File: squashfs-root/LiveOS/rootfs.img
  Size: 2411724800      Blocks: 4710400    IO Block: 4096   regular file
[...]

# dd if=/dev/zero of=squashfs-root/LiveOS/rootfs.img bs=1M seek=2300 count=700
700+0 records in
700+0 records out
734003200 bytes (734 MB, 700 MiB) copied, 0.308654 s, 2.4 GB/s

# ls -lh squashfs-root/LiveOS/rootfs.img
-rw-r--r--. 1 root root 3.0G Aug 14 20:17 squashfs-root/LiveOS/rootfs.img

# mount squashfs-root/LiveOS/rootfs.img /mnt/

# df -h /mnt/
Filesystem      Size  Used Avail Use% Mounted on
/dev/loop0      2.3G  1.5G  762M  67% /mnt

# xfs_growfs -d /mnt/
meta-data=/dev/loop0             isize=512    agcount=4, agsize=147200 blks
         =                       sectsz=512   attr=2, projid32bit=1
         =                       crc=1        finobt=1, sparse=1, rmapbt=0
         =                       reflink=1
data     =                       bsize=4096   blocks=588800, imaxpct=25
         =                       sunit=0      swidth=0 blks
naming   =version 2              bsize=4096   ascii-ci=0, ftype=1
log      =internal log           bsize=4096   blocks=2560, version=2
         =                       sectsz=512   sunit=0 blks, lazy-count=1
realtime =none                   extsz=4096   blocks=0, rtextents=0
data blocks changed from 588800 to 768000

# df -h /mnt/
Filesystem      Size  Used Avail Use% Mounted on
/dev/loop0      3.0G  1.5G  1.5G  52% /mnt

# umount /mnt
```

File: roles/addons/diskless/files/disklessset.py [hunk 2]
```diff
@@ -349,10 +349,10 @@ def generate_ipxe_boot_file(image_type, image_name, image_kernel, image_initramf
...
             print(bcolors.OKBLUE+'[INFO] Installing system into image.'+bcolors.ENDC)
             if selected_livenet_type == '3':
-                os.system('dnf install -y iproute procps-ng openssh-server  --installroot=/mnt/ --exclude glibc-all-langpacks --exclude cracklib-dicts --exclude grubby --exclude libxkbcommon --exclude pinentry --exclude python3-unbound --exclude unbound-libs --exclude xkeyboard-config --exclude trousers --exclude diffutils --exclude gnupg2-smime --exclude openssl-pkcs11 --exclude rpm-plugin-systemd-inhibit --exclude shared-mime-info --exclude glibc-langpack-* --setopt=module_platform_id=platform:el8 --nobest')
+                os.system('dnf install -y iproute procps-ng openssh-server  --installroot=/mnt/ --exclude glibc-all-langpacks --exclude cracklib-dicts --exclude grubby --exclude libxkbcommon --exclude pinentry --exclude python3-unbound --exclude unbound-libs --exclude xkeyboard-config --exclude trousers --exclude diffutils --exclude gnupg2-smime --exclude openssl-pkcs11 --exclude rpm-plugin-systemd-inhibit --exclude shared-mime-info --exclude glibc-langpack-* --setopt=module_platform_id=platform:el8 --releasever=8 --nobest')
```
Comment (btravouillon) [inline roles/addons/diskless/files/disklessset.py:352]: We don't support `$releasever` anymore in the stack.

File: roles/addons/diskless/files/disklessset.py [hunk 3]
```diff
@@ -349,10 +349,10 @@ def generate_ipxe_boot_file(image_type, image_name, image_kernel, image_initramf
...
+                os.system('dnf install -y iproute procps-ng openssh-server NetworkManager  --installroot=/mnt/ --exclude glibc-all-langpacks --exclude cracklib-dicts --exclude grubby --exclude libxkbcommon --exclude pinentry --exclude python3-unbound --exclude unbound-libs --exclude xkeyboard-config --exclude trousers --exclude diffutils --exclude gnupg2-smime --exclude openssl-pkcs11 --exclude rpm-plugin-systemd-inhibit --exclude shared-mime-info --exclude glibc-langpack-* --setopt=module_platform_id=platform:el8 --releasever=8 --nobest')
+                os.system('dnf install -y dnf  --installroot=/mnt/ --exclude glibc-all-langpacks --exclude cracklib-dicts --exclude grubby --exclude libxkbcommon --exclude pinentry --exclude python3-unbound --exclude unbound-libs --exclude xkeyboard-config --exclude trousers --exclude diffutils --exclude gnupg2-smime --exclude openssl-pkcs11 --exclude rpm-plugin-systemd-inhibit --exclude shared-mime-info --exclude glibc-langpack-* --setopt=module_platform_id=platform:el8 --releasever=8 --nobest')
```
Comment (btravouillon) [inline roles/addons/diskless/files/disklessset.py:355]: Same here, no `$releasever`.

File: roles/addons/diskless/files/disklessset.py [hunk 4]
```diff
@@ -559,4 +560,189 @@ def generate_ipxe_boot_file(image_type, image_name, image_kernel, image_initramf
             print(e)
             raise
 
+
+    elif sub_main_action == '5':
```
Comment (btravouillon) [inline roles/addons/diskless/files/disklessset.py:564]: Can you move this block between sub_main_actions 4 and 6?

File: roles/addons/diskless/vars/RedHat.yml [hunk 5]
```diff
@@ -1,2 +1,7 @@
 ---
 diskless_apache_user: apache
+diskless_packages_to_install: 
+  - httpd
+  - clustershell
```
Comment (btravouillon) [inline roles/addons/diskless/vars/RedHat.yml:5]: Is clustershell required or python3-clustershell is enough?

File: roles/addons/diskless/vars/RedHat.yml [hunk 6]
```diff
@@ -1,2 +1,7 @@
...
+  - python3-clustershell
```
Comment (btravouillon) [inline roles/addons/diskless/vars/RedHat.yml:6]: Might not be compatible with EL7.

File: roles/addons/diskless/readme.rst [hunk 7]
```diff
@@ -186,6 +187,66 @@ The -n parameter can be a nodeset.
...
+* The --skip-tags identify prevent hostname and static ip to be set, since the image should be generic for multiple hosts.
+
+Before closing, also remember to clean dnf cache into the image chroot to save space.
+
+Now, using df command, check used space of the image, to resize it later if whished.
+
+Using disklessset now, choose option 2 to unmount the image and squashfs it again.
+
+It is possible now to use the tool to resize image, to reduce it to the desired value (to save ram on target host).
+Always keep at least 100mb in / for temporary files and few logs generated during run.
```
Comment (btravouillon) [inline roles/addons/diskless/readme.rst:248]:
```suggestion
Always keep at least 100MB in / for temporary files and few logs generated during run.
```
Open question: is there a way to configure log_client to send logs to the log server only and not keep a copy on the node? (or only last logs with logrotate?)

File: roles/addons/diskless/vars/RedHat.yml (hunk 5 above)
Comment (oxedions) [reply roles/addons/diskless/vars/RedHat.yml]: python3-clustershell is enough. Fixed.

File: roles/addons/diskless/vars/RedHat.yml (hunk 6 above)
Comment (oxedions) [reply roles/addons/diskless/vars/RedHat.yml]: Splited for both RHEL

File: roles/core/nic/tasks/main.yml [hunk 8]
```diff
@@ -25,3 +25,5 @@
   when:
     - ( item.ip4 is defined and item.ip4 is not none ) or ( item.type is defined and item.type == 'bond-slave' )
     - item.interface is defined and item.interface is not none
+  tags:
+    - identify
```
[the remark this answers is not in the document]
Comment (oxedions) [reply roles/core/nic/tasks/main.yml:29]: Yes. The identify tag allows playbooks to be run on diskless mounted images. This change is related to a diskless need.

File: roles/core/set_hostname/tasks/main.yml [hunk 9]
```diff
@@ -2,3 +2,5 @@
 - name: "hostname █ Set {{ inventory_hostname }} hostname"
   hostname:
     name: "{{ inventory_hostname }}"
+  tags:
+    - identify
```
[the remark this answers is not in the document]
Comment (oxedions) [reply roles/core/set_hostname/tasks/main.yml:6]: Same

File: roles/addons/diskless/readme.rst (hunk 7 above)
Comment (oxedions) [reply roles/addons/diskless/readme.rst]: I think yes, I need to have a look :-)

File: roles/addons/diskless/readme.rst [hunk 10]
```diff
@@ -186,6 +187,66 @@ The -n parameter can be a nodeset.
       - pxe_stack
       - diskless
 
+Customizing Livenet image
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+It is assume here that image name used is *space_image*.
+
+The disklessset tool allows to customize livenet images before booting them, 
+and especially to execute playbooks inside images.
```
[the remark this answers is not in the document]
Comment (oxedions) [reply roles/addons/diskless/readme.rst:196]: fixed

File: roles/addons/diskless/files/disklessset.py [hunk 11]
```diff
@@ -601,6 +602,218 @@ def generate_ipxe_boot_file(image_type, image_name, image_kernel, image_initramf
...
+                os.makedirs(image_working_directory + '_copy')
+            except FileExistsError:
+                print(bcolors.WARNING + '[WARNING] The directory ' + image_working_directory + '_copy' + ' already exists. Cleaning.' + bcolors.ENDC)
+                shutil.rmtree(image_working_directory + '_copy')
+                os.makedirs(image_working_directory + '_copy')
+            except OSError:
+                print(bcolors.FAIL + '[ERROR] Cannot create directory ' + image_working_directory + '_copy' + bcolors.ENDC)
+
+            try:
+                os.makedirs(image_working_directory + '_copy/squashfs-root/LiveOS/')
```
[the remark this answers is not in the document]
Comment (oxedions) [reply roles/addons/diskless/files/disklessset.py:749]: Fixed

**Pull request merged**